transports/janus_pfunix.c

```diff
@@ -0,0 +1,525 @@
+int janus_pfunix_send_message(void *transport, void *request_id, gboolean admin, json_t *message);
+void janus_pfunix_session_created(void *transport, guint64 session_id);
+void janus_pfunix_session_over(void *transport, guint64 session_id, gboolean timeout);
+
+
+/* Transport setup */
+static janus_transport janus_pfunix_transport =
+	JANUS_TRANSPORT_INIT (
+		.init = janus_pfunix_init,
+		.destroy = janus_pfunix_destroy,
+
+		.get_api_compatibility = janus_pfunix_get_api_compatibility,
+		.get_version = janus_pfunix_get_version,
+		.get_version_string = janus_pfunix_get_version_string,
+		.get_description = janus_pfunix_get_description,
+		.get_name = janus_pfunix_get_name,
+		.get_author = janus_pfunix_get_author,
+		.get_package = janus_pfunix_get_package,
+
+		.is_janus_api_enabled = janus_pfunix_is_janus_api_enabled,
+		.is_admin_api_enabled = janus_pfunix_is_admin_api_enabled,
+
+		.send_message = janus_pfunix_send_message,
+		.session_created = janus_pfunix_session_created,
+		.session_over = janus_pfunix_session_over,
+	);
+
+/* Transport creator */
+janus_transport *create(void) {
+	JANUS_LOG(LOG_VERB, "%s created!\n", JANUS_PFUNIX_NAME);
+	return &janus_pfunix_transport;
+}
+
+
+/* Useful stuff */
+static gint initialized = 0, stopping = 0;
+static janus_transport_callbacks *gateway = NULL;
+
+#define BUFFER_SIZE		8192
+
+
+/* Unix Sockets server thread */
+static GThread *pfunix_thread = NULL;
+void *janus_pfunix_thread(void *data);
+
+/* Unix Sockets servers */
+static int pfd = -1, admin_pfd = -1;
+/* Socket pair to notify about the need for outgoing data */
+static int write_fd[2];
+
+/* Unix Sockets client session */
+typedef struct janus_pfunix_client {
+	int fd;						/* Client socket */
+	gboolean admin;				/* Whether this client is for the Admin or Janus API */
+	GAsyncQueue *messages;		/* Queue of outgoing messages to push */
+	gboolean session_timeout;	/* Whether a Janus session timeout occurred in the core */
+} janus_pfunix_client;
+static GHashTable *clients = NULL, *clients_by_fd = NULL;
+static janus_mutex clients_mutex;
+
+
+/* Transport implementation */
+int janus_pfunix_init(janus_transport_callbacks *callback, const char *config_path) {
+	if(g_atomic_int_get(&stopping)) {
+		/* Still stopping from before */
+		return -1;
+	}
+	if(callback == NULL || config_path == NULL) {
+		/* Invalid arguments */
+		return -1;
+	}
+
+	/* This is the callback we'll need to invoke to contact the gateway */
+	gateway = callback;
+
+	/* Read configuration */
+	char filename[255];
+	g_snprintf(filename, 255, "%s/%s.cfg", config_path, JANUS_PFUNIX_PACKAGE);
+	JANUS_LOG(LOG_VERB, "Configuration file: %s\n", filename);
+	janus_config *config = janus_config_parse(filename);
+	if(config != NULL) {
+		/* Handle configuration */
+		janus_config_print(config);
+
+		/* First of all Initialize the socketpair for writeable notifications */
+		if(socketpair(PF_LOCAL, SOCK_STREAM, 0, write_fd) < 0) {
```

Comment:
these also need to be non-blocking, though we might be able to get away with it.  Or maybe we can use a `pipe` and set the read end in non-blocking mode. I'll check what's best.